homeassistant/components/upb/__init__.py

```diff
@@ -0,0 +1,123 @@
+"""Support the UPB PIM."""
+import asyncio
+import logging
+
+import upb_lib
+
+from homeassistant.const import CONF_FILE_PATH, CONF_HOST
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DOMAIN
+
+UPB_PLATFORMS = ["light"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, hass_config: ConfigType) -> bool:
+    """Set up the UPB platform."""
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up a new config_entry for UPB PIM."""
+
+    url = config_entry.data[CONF_HOST]
+    file = config_entry.data[CONF_FILE_PATH]
+
+    upb = upb_lib.UpbPim({"url": url, "UPStartExportFile": file})
+    upb.connect()
+    hass.data[DOMAIN] = {"upb": upb}
+
+    for component in UPB_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload the config_entry."""
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in UPB_PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        upb = hass.data[DOMAIN]["upb"]
+        upb.disconnect()
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
+class UpbEntity(Entity):
+    """Base class for all UPB entities."""
+
+    def __init__(self, element, upb):
+        """Initialize the base of all UPB devices."""
+        self._upb = upb
+        self._element = element
+        self._unique_id = f"{self.__class__.__name__.lower()}_{element.index}"
```

Comment:
Please check the unique id requirements here 

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

Reply:
There is no other option that I know of. Similar to the ElkM1 in that sense. This is a serial connected wall wart that has not serial number or any other unique identifier.

Does using the `entry_id` mean that when the integration is dropped and re-added that the `unique_id` will change? I don't think that matters to any automations as the `entity_id` will remain the same. Is that correct?

Can you point me to an integration that is using `entry_id` so I can see how that is done?

Reply:
I did this in `nut` when there is no serial number from the ups.  In hindsight I probably should have ignored the serial number after I found out how unreliable ups manufacturers at reporting it.